internal/xds/balancer/outlierdetection/balancer_test.go

```diff
@@ -1027,253 +1028,348 @@ func (s) TestDurationOfInterval(t *testing.T) {
+			lbConfig: LBConfig{
+				Interval:           math.MaxInt64, // so the interval will never run unless called manually in test.
+				BaseEjectionTime:   iserviceconfig.Duration(30 * time.Second),
+				MaxEjectionTime:    iserviceconfig.Duration(300 * time.Second),
+				MaxEjectionPercent: 10,
+				FailurePercentageEjection: &FailurePercentageEjection{
+					Threshold:             50,
+					EnforcementPercentage: 0,
+					MinimumHosts:          3,
+					RequestVolume:         3,
+				},
+				ChildPolicy: &iserviceconfig.BalancerConfig{
+					Name:   "three_upstreams_one_failure_no_ejection_enforcement_perc",
+					Config: emptyChildConfig{},
+				},
 			},
-			ChildPolicy: &iserviceconfig.BalancerConfig{
-				Name:   t.Name(),
-				Config: emptyChildConfig{},
+			numberOfConns: 3,
+			wantFailures:  1,
+			wantEjections: 0,
+		},
+		{
+			name: "three_upstreams_one_failure_no_ejection_max_ejection_perc",
+			lbConfig: LBConfig{
+				Interval:           math.MaxInt64, // so the interval will never run unless called manually in test.
+				BaseEjectionTime:   iserviceconfig.Duration(30 * time.Second),
+				MaxEjectionTime:    iserviceconfig.Duration(300 * time.Second),
+				MaxEjectionPercent: 0,
+				FailurePercentageEjection: &FailurePercentageEjection{
+					Threshold:             50,
+					EnforcementPercentage: 100,
+					MinimumHosts:          3,
+					RequestVolume:         3,
+				},
+				ChildPolicy: &iserviceconfig.BalancerConfig{
+					Name:   "three_upstreams_one_failure_no_ejection_max_ejection_perc",
+					Config: emptyChildConfig{},
+				},
 			},
+			numberOfConns: 3,
+			wantFailures:  1,
+			wantEjections: 0,
 		},
-	})
-
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-
-	// Transition the SubConns to READY so that they can register health
-	// listeners.
-	for range 3 {
-		select {
-		case <-ctx.Done():
-			t.Fatalf("Timed out waiting for creation of new SubConn.")
-		case sc := <-tcc.NewSubConnCh:
-			sc.UpdateState(balancer.SubConnState{ConnectivityState: connectivity.Connecting})
-			sc.UpdateState(balancer.SubConnState{ConnectivityState: connectivity.Ready})
-		}
 	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			scsCh := testutils.NewChannel()
+			connectivityCh := make(chan struct{})
+			var allSubConns = make([]balancer.SubConn, test.numberOfConns)
+			stub.Register(test.name, stub.BalancerFuncs{
+				UpdateClientConnState: func(bd *stub.BalancerData, _ balancer.ClientConnState) error {
+					for i := range test.numberOfConns {
+						scw, err := bd.ClientConn.NewSubConn([]resolver.Address{{Addr: fmt.Sprintf("address%d", i+1)}}, balancer.NewSubConnOptions{
+							StateListener: func(state balancer.SubConnState) {
+								if state.ConnectivityState == connectivity.Ready {
+									connectivityCh <- struct{}{}
+								}
+							},
+						})
+						if err != nil {
+							t.Errorf("error in od.NewSubConn call: %v", err)
```

Comment:
Nit: Could you please improve this error message to be something like
`t.Errorf("NewSubConn(%q) failed: %v", addr, err)`

You would have to pull the address out of the endpoints from the `balancer.ClientConnState` passed to this method.

Reply:
fixed in 32d9eb3ea51d51cdcf5c8ae1477f6c435ed878bd, ended up using `balancer.ClientConnState` to get the endpoints so it was nice